keyboards/al1/keymaps/via/keymap.c

```diff
@@ -0,0 +1,46 @@
+/* Copyright 2018 MechMerlin
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "al1.h"
```

Comment:
```suggestion
#include QMK_KEYBOARD_H
```

Reply:
That's odd. It compiled with an older version of qmk. Now that I have updated/rebuilt my environment I get the same size error. I have applied all of the space saving tricks I know of, but it is still too big. However, it does compile when I change to "VIA_ENABLE = no", and with about 50% free space! Does enabling VIA really take up that much room, or am I missing something?

I will make the suggested edits and commit the updates once I have a working keymap.

Also, sorry if I'm doing something wrong/out of order here, this is my first time creating a PR, and I'm new to github in general. Thanks!

Reply:
The issue isn't with flash space, it's about EEPROM. The way VIA works is that on first run, the keymap stored in flash is copied to the EEPROM, because on AVR devices flash is read-only at runtime. However the EEPROM size is much smaller than that of flash (1024 bytes on ATmega32U4, refer to datasheet), so there are only so many layers you can fit. The general formula is `2 * MATRIX_ROWS * MATRIX_COLS * layers`, less at least 100 bytes for VIA macro storage, and ~34 bytes for "core" QMK EEPROM stuff.

Reply:
I see. Since the only variable in that equation that I have control over is the number of layers, I went ahead and tried to compile with fewer and fewer layers in the keymap without luck. Does this mean that the AL1 is impossible to port to VIA simply because there isn't enough EEPROM?

Reply:
Not impossible, actually it seems the culprit is due to `MATRIX_ROWS` and `MATRIX_COLS` not matching the real matrix size, which will obviously throw off the calculation. I've already fixed this in the PR I linked below.

`2 * 6 * 16 * 4` comes out to 768 bytes, which is about average for a board of this size. If it were too large, you would need to reduce the layer count VIA assumes (it does not look at the size of the keymap itself) by adding `#define DYNAMIC_KEYMAP_LAYER_COUNT 3` in config.h, but this should not be necessary here.

Reply:
Using your PR #16331 files I was able to compile and flash the via firmware! But my AL1 is not responsive anymore... the numlock indicator led still lights up when plugged in tho. did I brick this pcb?

Reply:
I think I have the enable logic inverted in the 154 driver. Try changing lines 36 and 41 to `writePinHigh`, and 47 and 50 to `!enabled`.

Reply:
That did the trick! Everything seems to be working perfectly. So, does this mean that I should wait until your PR #16331 gets approved before I commit any revisions to this PR (since it only works with your driver/custom matrix)? 

Then we should be able to create a new PR in the VIA github for an AL1 json file, and THEN I'll be able to program my AL1 with VIA! ...right?

Reply:
My PR is targeting the develop branch, so you would have to wait for the next breaking changes cycle in about 3 months. However, you can just fix `MATRIX_ROWS` and `MATRIX_COLS` here, as that's really all that needs to be done to solve the compile error, and I can rebase later.